lib/response.js

```diff
@@ -404,68 +364,75 @@ function __send(maybeCode, maybeBody, maybeHeaders, maybeCallback, format) {
-            return _flush();
+    // At this point we know we have a body that needs to be formatted, so lets
+    // derive the formatter based on the response object's properties
+
+    // _formatterError is used to handle any case where we were unable to
+    // properly format the provided body
+    function _formatterError(err) {
+        // If the user provided a non-success error code, we don't want to mess
+        // with it since their error is probably more important than our
+        // inability to format their message.
+        if (self.statusCode >= 200 && self.statusCode < 300) {
+            self.statusCode = err.statusCode;
         }
 
-        // if we have formatter, happy path.
-        var asyncFormat = (formatter && formatter.length === 4) ? true : false;
+        log.warn({
+            req: self.req,
+            err: err
+        }, 'error retrieving formatter');
 
-        if (asyncFormat === true) {
+        return _flush();
+    }
 
-            assert.func(callback, 'async formatter for ' + contentType +
-                                  ' requires callback to res.send()');
+    var formatter;
+    var type = self.contentType || self.getHeader('Content-Type');
+
+    // Check to see if we can find a valid formatter
+    if (!type && !self.req.accepts(self.acceptable)) {
+        return _formatterError(new errors.NotAcceptableError({
+            message: 'could not find suitable formatter'
+        }));
+    }
 
-            // if async formatter error, propagate error back up to
-            // res.send() caller, most likely a handler.
-            return formatter.call(self, self.req, self, body,
-            function _formatDone(formatErr, formattedBody) {
+    // Derive type if not provided by the user
+    if (!type) {
+        type = self.req.accepts(self.acceptable);
+    }
 
-                if (formatErr) {
+    type = type.split(';')[0];
 
-                    return callback(new errors.FormatterError(formatErr, {
-                        message: 'unable to format response for ' +
-                                    self.header('content-type'),
-                        context: {
-                            rawBody: body
-                        }
-                    }));
-                }
-                return _flush(formattedBody);
-            });
-        }
-        // for sync formatters, invoke formatter and send response body.
-        else {
-            _flush(formatter.call(self, self.req, self, body), body);
-            // users can always opt to pass in next, even when formatter is not
-            // async. invoke callback in that case. return null when no
-            // callback because eslint wants consistent-return.
-            return (callback) ? callback() : null;
-        }
-    });
+    if (!self.formatters[type] && type.indexOf('/') === -1) {
+        type = mime.lookup(type);
+    }
+
+    if (!self.formatters[type] && self.acceptable.indexOf(type) === -1) {
```

Comment:
Can we add a code comment explaining why we set type to `application/octet-stream`